### PR TITLE
agent method to GET message

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Messaging/Agent.hs [hunk 1]
```diff
@@ -345,7 +355,7 @@ processConfirmation c rq@RcvQueue {e2ePrivKey} SMPConfirmation {senderKey, e2ePu
 
 -- | Subscribe to receive connection messages (SUB command) in Reader monad
 subscribeConnection' :: forall m. AgentMonad m => AgentClient -> ConnId -> m ()
-subscribeConnection' c connId =
+subscribeConnection' c connId = do
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent.hs:358]:
```suggestion
subscribeConnection' c connId =
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 2]
```diff
@@ -113,6 +114,7 @@ data AgentClient = AgentClient
     connMsgsQueued :: TMap ConnId Bool,
     smpQueueMsgQueues :: TMap (ConnId, SMPServer, SMP.SenderId) (TQueue InternalId),
     smpQueueMsgDeliveries :: TMap (ConnId, SMPServer, SMP.SenderId) (Async ()),
+    getMsgLocks :: TMap (ConnId, SMPServer, SMP.RecipientId) (TMVar ()),
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:117]:
```suggestion
    msgLocks :: TMap (ConnId, SMPServer, SMP.RecipientId) (TMVar ()),
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 3]
```diff
@@ -137,11 +139,12 @@ newAgentClient InitialAgentServers {smp, ntf} agentEnv = do
   connMsgsQueued <- TM.empty
   smpQueueMsgQueues <- TM.empty
   smpQueueMsgDeliveries <- TM.empty
+  getMsgLocks <- TM.empty
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:142]:
```suggestion
  msgLocks <- TM.empty
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 4]
```diff
@@ -137,11 +139,12 @@ newAgentClient InitialAgentServers {smp, ntf} agentEnv = do
...
   reconnections <- newTVar []
   asyncClients <- newTVar []
   clientId <- stateTVar (clientCounter agentEnv) $ \i -> (i + 1, i + 1)
   lock <- newTMVar ()
-  return AgentClient {active, rcvQ, subQ, msgQ, smpServers, smpClients, ntfServers, ntfClients, subscrSrvrs, pendingSubscrSrvrs, subscrConns, connMsgsQueued, smpQueueMsgQueues, smpQueueMsgDeliveries, reconnections, asyncClients, clientId, agentEnv, lock}
+  return AgentClient {active, rcvQ, subQ, msgQ, smpServers, smpClients, ntfServers, ntfClients, subscrSrvrs, pendingSubscrSrvrs, subscrConns, connMsgsQueued, smpQueueMsgQueues, smpQueueMsgDeliveries, getMsgLocks, reconnections, asyncClients, clientId, agentEnv, lock}
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:147]:
```suggestion
  return AgentClient {active, rcvQ, subQ, msgQ, smpServers, smpClients, ntfServers, ntfClients, subscrSrvrs, pendingSubscrSrvrs, subscrConns, connMsgsQueued, smpQueueMsgQueues, smpQueueMsgDeliveries, msgLocks, reconnections, asyncClients, clientId, agentEnv, lock}
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 5]
```diff
@@ -423,6 +426,7 @@ newRcvQueue_ a c srv = do
 
 subscribeQueue :: AgentMonad m => AgentClient -> RcvQueue -> ConnId -> m ()
 subscribeQueue c rq@RcvQueue {server, rcvPrivateKey, rcvId} connId = do
+  whenM (atomically . TM.member (connId, server, rcvId) $ getMsgLocks c) . throwError $ CMD PROHIBITED
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:429]:
```suggestion
  whenM (atomically . TM.member (connId, server, rcvId) $ msgLocks c) . throwError $ CMD PROHIBITED
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 6]
```diff
@@ -501,6 +505,23 @@ sendInvitation c (Compatible SMPQueueInfo {smpServer, senderId, dhPublicKey}) co
       agentCbEncryptOnce dhPublicKey . smpEncode $
         SMP.ClientMessage SMP.PHEmpty $ smpEncode agentEnvelope
 
+getQueueMessage :: AgentMonad m => AgentClient -> RcvQueue -> ConnId -> m (Maybe (MsgId, MsgFlags))
+getQueueMessage c@AgentClient {getMsgLocks} RcvQueue {server, rcvId, rcvPrivateKey} connId =
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:509]:
```suggestion
getQueueMessage c@AgentClient {msgLocks} RcvQueue {server, rcvId, rcvPrivateKey} connId =
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 7]
```diff
@@ -501,6 +505,23 @@ sendInvitation c (Compatible SMPQueueInfo {smpServer, senderId, dhPublicKey}) co
...
+  E.bracket (atomically createTakeLock) (atomically . (`putTMVar` ())) $ \_ ->
+    withLogClient c server rcvId "GET" $ \smp ->
+      getSMPMessage smp rcvPrivateKey rcvId
+  where
+    k = (connId, server, rcvId)
+    createTakeLock = do
+      l <- TM.lookup k getMsgLocks >>= maybe newLock pure
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:516]:
```suggestion
      l <- TM.lookup k msgLocks >>= maybe newLock pure
```

File: src/Simplex/Messaging/Agent/Client.hs [hunk 8]
```diff
@@ -501,6 +505,23 @@ sendInvitation c (Compatible SMPQueueInfo {smpServer, senderId, dhPublicKey}) co
...
+      takeTMVar l
+      pure l
+      where
+        newLock = do
+          l <- newTMVar ()
+          TM.insert k l getMsgLocks
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Agent/Client.hs:522]:
```suggestion
          TM.insert k l msgLocks
```

File: src/Simplex/Messaging/Client.hs [hunk 9]
```diff
@@ -287,6 +288,18 @@ subscribeSMPQueue c@ProtocolClient {protocolServer, sessionId, msgQ} rpKey rId =
       lift . atomically $ mapM_ (`writeTBQueue` (protocolServer, sessionId, rId, cmd)) msgQ
     _ -> throwE PCEUnexpectedResponse
 
+-- | Get message from SMP queue. The server returns ERR PROHIBITED if a client uses SUB and GET via the same transport connection for the same queue
+--
+-- https://github.covm/simplex-chat/simplexmq/blob/master/protocol/simplex-messaging.md#receive-a-message-from-the-queue
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Client.hs:293]: the link doesn't seem to work.. I've checked this and "notifications" branches as well, maybe I'm missing from which branch it's going to be merged

File: src/Simplex/Messaging/Agent/Client.hs (hunk 3 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Agent/Client.hs:142]: no, they are GET locks :)

File: src/Simplex/Messaging/Client.hs (hunk 9 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Client.hs:293]: need to be added

**Pull request merged**